### DISCUSSION TITLE
Future upgrades when forking repo?

### DISCUSSION THREAD

**Discussion by njackson1582** [microsoft/AL-Go#122] [Q&A] [answered]
The [AL-Go-PTE template](https://github.com/microsoft/AL-Go-PTE) is a great start for everything we want to do regarding DevOps, but there are some specific features we'd like to have in a standard PTE repo (code owners, object ranges, etc.). As GitHub [doesn't allow for template variables](https://github.com/isaacs/github/issues/1716) (yet), I'm thinking something simple like a PowerShell script would be nice to have in each PTE that the developer can run which will programmatically make configurations to the repo as needed. However, to customize the template, I'd have to fork it. 

If I fork the template to add my own PowerShell script, would I still be able to receive regular updates based upon the original AL-Go-PTE repository, ideally with the `Update AL-Go System Files` action? Or would I be stuck performing manual updates on each PTE repo?

**Comment by freddydk** [comment, marked as answer]
AL-Go was created with this scenario in mind as well.
It is kind of explained in this document: https://github.com/microsoft/AL-Go/blob/main/Scenarios/Contributing.md
You fork the AL-Go repo to your org - and then you run the Deploy action in your fork of the AL-Go repo.
Then you will have 3 repos:
- https://github.com/yourorg/AL-Go-PTE
- https://github.com/yourorg/AL-Go-Actions
- https://github.com/yourorg/AL-Go-AppSource

Remember to set PTE and AppSource app repos as templates.
Then you use your own PTE or AppSource app repo as template or you run Update AL-Go System files specifying your PTE or AppSource app as template repo.
Now your repo is using your PTE or AppSource app template.
If you make changes to "your three repos", you can run Collect in your fork of the AL-Go repo to collect the changes.
Now you can Fetch Upstream changes from microsoft/AL-Go to your fork and resolve conflicts if any.
And then re-run Deploy - potentially adding a new version number of incompatibilities might have been implemented (very rare)

So, you can kind of see the AL-Go repo as "the source" and the 3 repos as "the installation of AL-Go"

If your changes are good and useful for other partners - you can even submit them as contributions
Note that AL-Go forks and template repos must be public - I don't think the system supports private repos for this, but then again - your secrets are not in the AL-Go repos.

**Reply by njackson1582** [reply]
Perfect, this explains well the 3-repo model that I was a bit confused on. Thanks!

**Comment by freddydk** [comment]
Since this - a lot of changes have been made to AL-Go and you can now use a specific branch of your AL-Go repository as your template repository - but... - we are also working on creating a feature called indirect templates, which allows you to define your own template based on AL-Go.

**Reply by dannoe** [reply]
Sorry for hijacking this discussion for my question, but its similar.

@freddydk Would indirect templates allow us to add a custom build steps to our pipeline?

Background:
We are still using the original azure devops pipeline and have customized some things (mostly powershell scripts that do something to the AL code before building it). We are thinking about moving to AL-Go. Would our scenario be possible with indirect templates or is this already possible somehow? If I understood AL-Go for github correctly, our custom scripts will be overwritten everytime after updating the AL-Go system files.

PS: Please let me know if I should open a separate discussion for this topic.

**Reply by dannoe** [reply]
I guess I found the solution in this discussion: https://github.com/microsoft/AL-Go/discussions/693
It is possible to add custom jobs to the workflows and they should be maintained on updating the AL-GO system files.
